packages/core-ethereum/crates/core-ethereum-db/src/db.rs

```diff
@@ -0,0 +1,626 @@
+        let _ = self.db.set(key, &commitment).await?;
+        Ok(())
+    }
+
+    async fn get_latest_block_number(&self) -> Result<u32> {
+        let key = utils_db::db::Key::new_from_str(LATEST_BLOCK_NUMBER_KEY)?;
+        self.db.get_or_none::<u32>(key).await.map(|v| v.unwrap_or(0))
+    }
+
+    async fn update_latest_block_number(&mut self, number: u32) -> Result<()> {
+        let key = utils_db::db::Key::new_from_str(LATEST_BLOCK_NUMBER_KEY)?;
+        let _ = self.db.set(key, &number).await?;
+        Ok(())
+    }
+
+    async fn get_latest_confirmed_snapshot(&self) -> Result<Option<Snapshot>> {
+        let key = utils_db::db::Key::new_from_str(LATEST_CONFIRMED_SNAPSHOT_KEY)?;
+        self.db.get_or_none::<Snapshot>(key).await
+    }
+
+    async fn get_channel(&self, channel: &Hash) -> Result<Option<ChannelEntry>> {
+        let key = utils_db::db::Key::new_with_prefix(channel, CHANNEL_PREFIX)?;
+        self.db.get_or_none::<ChannelEntry>(key).await
+    }
+
+    async fn get_channels(&self) -> Result<Vec<ChannelEntry>> {
+        self.db
+            .get_more::<ChannelEntry>(Box::from(CHANNEL_PREFIX.as_bytes()), Hash::size(), &|_| true)
+            .await
+    }
+
+    async fn get_channels_open(&self) -> Result<Vec<ChannelEntry>> {
+        Ok(self
+            .db
+            .get_more::<ChannelEntry>(Box::from(CHANNEL_PREFIX.as_bytes()), Hash::size(), &|_| true)
+            .await?
+            .into_iter()
+            .filter(move |x| x.status == ChannelStatus::Open)
+            .collect())
+    }
+
+    async fn update_channel_and_snapshot(
+        &mut self,
+        channel_id: &Hash,
+        channel: ChannelEntry,
+        snapshot: Snapshot,
+    ) -> Result<()> {
+        let channel_key = utils_db::db::Key::new_with_prefix(channel_id, CHANNEL_PREFIX)?;
+        let snapshot_key = utils_db::db::Key::new_from_str(LATEST_CONFIRMED_SNAPSHOT_KEY)?;
+
+        let mut batch_ops = utils_db::db::Batch::new();
+        batch_ops.put(channel_key, channel);
+        batch_ops.put(snapshot_key, snapshot);
+
+        self.db.batch(batch_ops, true).await
+    }
+
+    async fn get_account(&self, address: &Address) -> Result<Option<AccountEntry>> {
+        let key = utils_db::db::Key::new_with_prefix(address, ACCOUNT_PREFIX)?;
+        self.db.get_or_none::<AccountEntry>(key).await
+    }
+
+    async fn update_account_and_snapshot(&mut self, account: AccountEntry, snapshot: Snapshot) -> Result<()> {
+        let address_key = utils_db::db::Key::new_with_prefix(&account.get_address(), ACCOUNT_PREFIX)?;
+        let snapshot_key = utils_db::db::Key::new_from_str(LATEST_CONFIRMED_SNAPSHOT_KEY)?;
+
+        let mut batch_ops = utils_db::db::Batch::new();
+        batch_ops.put(address_key, account);
+        batch_ops.put(snapshot_key, snapshot);
+
+        self.db.batch(batch_ops, true).await
+    }
+
+    async fn get_accounts(&self, address: Address) -> Result<Vec<AccountEntry>> {
```

Comment:
Shouldn't the `address` argument be used to filter ?

Reply:
There are lifetime issues with these arguments, so I extracted the filtered option into a different callback, but will double check. Some of the callback values can be passed by reference as well. Will do a final sweep today.